examples/examples/jsonrpsee_server_low_level_api.rs

```diff
@@ -0,0 +1,286 @@
+
+	let stop_handle = StopHandle::new(rx);
+	let server_handle = ServerHandle::new(tx);
+	let methods = ().into_rpc();
+	let conn_guard = ConnectionGuard::new(100);
+	let conn_id = Arc::new(AtomicU32::new(0));
+	let http_rate_limit = Arc::new(AsyncMutex::new(0));
+
+	// Blacklisted peers
+	let blacklisted_peers = Arc::new(Mutex::new(HashSet::new()));
+	let stop_handle2 = stop_handle.clone();
+
+	// And a MakeService to handle each connection...
+	let make_service = make_service_fn(move |conn: &AddrStream| {
+		// Connection state.
+		let conn_id = conn_id.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+		let remote_addr = conn.remote_addr();
+		let stop_handle = stop_handle2.clone();
+		let conn_guard = conn_guard.clone();
+		let methods = methods.clone();
+		let blacklisted_peers = blacklisted_peers.clone();
+		let http_rate_limit = http_rate_limit.clone();
+
+		async move {
+			let stop_handle = stop_handle.clone();
+			let conn_guard = conn_guard.clone();
+			let methods = methods.clone();
+			let stop_handle = stop_handle.clone();
+			let blacklisted_peers = blacklisted_peers.clone();
+			let http_rate_limit = http_rate_limit.clone();
+
+			Ok::<_, Infallible>(service_fn(move |req| {
+				// Connection number limit exceeded.
+				let Some(conn_permit) = conn_guard.try_acquire() else {
+					return async { Ok::<_, Infallible>(http::response::too_many_requests()) }.boxed();
+				};
+
+				// The IP addr was blacklisted.
+				if blacklisted_peers.lock().unwrap().get(&remote_addr.ip()).is_some() {
+					return async { Ok(http::response::denied()) }.boxed();
+				}
+
+				if ws::is_upgrade_request(&req) {
+					let methods = methods.clone();
+					let stop_handle = stop_handle.clone();
+					let blacklisted_peers = blacklisted_peers.clone();
+
+					let (tx, mut disconnect) = mpsc::channel(1);
+					let rpc_service = RpcServiceBuilder::new().layer_fn(move |service| DummyRateLimit {
+						service,
+						count: Default::default(),
+						state: tx.clone(),
+					});
+
+					let params = to_jsonrpsee_params(methods.clone(), stop_handle.clone(), conn_id, conn_permit);
+
+					// Establishes the websocket connection
+					// and if the `DummyRateLimit` middleware triggers the hard limit
+					// then the connection is closed i.e, the `conn_fut` is dropped.
+					async move {
+						match ws::connect(req, params, rpc_service).await {
+							Ok((rp, conn_fut)) => {
+								tokio::spawn(async move {
+									tokio::select! {
+										_ = conn_fut => (),
+										_ = disconnect.recv() => {
+											blacklisted_peers.lock().unwrap().insert(remote_addr.ip());
+										},
+									}
+								});
+								Ok(rp)
+							}
+							Err(rp) => Ok(rp),
+						}
+					}
+					.boxed()
+				} else if !ws::is_upgrade_request(&req) {
+					// In this example it doesn't make sense to disconnect the peer in the middleware rate limit
+					// is triggered as it will just reply to HTTP request i.e, the "mpsc::Sender disconnect"
+					// is not used.
```

Comment:
But this means that an IP address used for HTTP connections will never be blacklisted, unlike WS?

Reply:
It will be blacklisted but the channel is not used to "kill the connection" because for HTTP it's only one call to complete
and it makes no sense to be cancel a RPC on HTTP.

This example blacklisting per HTTP request won't work in practice because one would need a HashMap or something to just ban if the same IP peer is making many requests.

This will blacklist the request which fires trigger the rate limit but it could in fact be the first HTTP request it sends ^^

Reply:
Changed the code to avoid writing such weird comments